indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/Asn1Decoding.kt

```diff
@@ -0,0 +1,385 @@
+package at.asitplus.signum.indispensable.asn1.encoding
+
+import at.asitplus.catching
+import at.asitplus.signum.indispensable.asn1.*
+import at.asitplus.signum.indispensable.asn1.BERTags.BMP_STRING
+import at.asitplus.signum.indispensable.asn1.BERTags.IA5_STRING
+import at.asitplus.signum.indispensable.asn1.BERTags.NUMERIC_STRING
+import at.asitplus.signum.indispensable.asn1.BERTags.PRINTABLE_STRING
+import at.asitplus.signum.indispensable.asn1.BERTags.T61_STRING
+import at.asitplus.signum.indispensable.asn1.BERTags.UNIVERSAL_STRING
+import at.asitplus.signum.indispensable.asn1.BERTags.UTF8_STRING
+import at.asitplus.signum.indispensable.asn1.BERTags.VISIBLE_STRING
+import com.ionspin.kotlin.bignum.integer.BigInteger
+import com.ionspin.kotlin.bignum.integer.util.fromTwosComplementByteArray
+import kotlinx.datetime.Instant
+import kotlin.experimental.and
+
+
+/**
+ * Parses the provided [input] into a single [Asn1Element]. Consumes all Bytes and throws if more than one Asn.1 Structure was found or trailing bytes were detected
+ * @return the parsed [Asn1Element]
+ *
+ * @throws Asn1Exception on invalid input or if more than a single root structure was contained in the [input]
+ */
+@Throws(Asn1Exception::class)
+fun Asn1Element.Companion.parse(source: ByteIterator): Asn1Element = parseAll(source).let{
+    if (it.size != 1) throw Asn1StructuralException("Multiple ASN.1 structures found")
+    it.first()
+}
+/**
+ * Convenience wrapper around [parse], taking a [ByteArray] as [source]
+ * @see parse
+ */
+@Throws(Asn1Exception::class)
+fun Asn1Element.Companion.parse(source: ByteArray): Asn1Element = parse(source.iterator())
+
+/**
+ * Tries to parse the [input] into a list of [Asn1Element]s. Consumes all Bytes and throws if an invalid ASN.1 Structure is found at any point.
+ * @return the parsed elements
+ *
+ * @throws Asn1Exception on invalid input or if more than a single root structure was contained in the [input]
+ */
+@Throws(Asn1Exception::class)
+fun Asn1Element.Companion.parseAll(input: ByteIterator): List<Asn1Element> = Asn1Reader(input).doParseAll()
+
+
+/**
+ * Parses the first [Asn1Element] from [input].
+ * @return the parsed [Asn1Element]. Trailing byte are left untouched and can be consumed from [input] after parsing
+ *
+ * @throws Asn1Exception on invalid input or if more than a single root structure was contained in the [input]
+ */
+//this only makes sense until we switch to kotlinx.io
+@Throws(Asn1Exception::class)
+fun Asn1Element.Companion.parseFirst(input: ByteIterator): Asn1Element = Asn1Reader(input).doParseSingle()
+
+
+private class Asn1Reader(private val input: ByteIterator) {
```

Comment:
do we still need this class? couldn't these just be extension functions on `ByteIterator`?

Reply:
right!